api/v1_coins_post_redeem.go

```diff
@@ -0,0 +1,234 @@
+	if err != nil {
+		return "", fmt.Errorf("failed to get attestation bytes: %w", err)
+	}
+
+	hash := crypto.Keccak256(claimData)
+
+	signatureBytes, err := crypto.Sign(hash, claimAuthorityKey)
+	if err != nil {
+		return "", fmt.Errorf("failed to sign hash: %w", err)
+	}
+
+	return "0x" + hex.EncodeToString(signatureBytes), nil
+}
+
+// TODO: Implement/replace once core reward generation PR is merged
+func getClaimAuthority(mint solana.PublicKey) (*ecdsa.PrivateKey, error) {
+	return nil, nil
+}
+
+type CoinRewardParams struct {
+	Mint   string
+	Amount uint64
+}
+
+type RewardCodeRow struct {
+	Mint          string `db:"mint"`
+	RewardAddress string `db:"reward_address"`
+	Amount        uint64 `db:"amount"`
+	IsUsed        bool   `db:"is_used"`
+}
+
+func (app *ApiServer) v1CoinsPostRedeem(c *fiber.Ctx) error {
+	mintString := c.Params("mint")
+	if mintString == "" {
+		return fiber.NewError(fiber.StatusBadRequest, "Mint parameter is required")
+	}
+
+	// Read optional code from route params
+	redeemCode := c.Params("code")
+
+	myId := app.getMyId(c)
+	if myId == 0 {
+		return fiber.NewError(fiber.StatusBadRequest, "Missing user ID")
+	}
+
+	var userWalletAddress string
+	var userHandle string
+	err := app.pool.QueryRow(c.Context(), `SELECT wallet, handle FROM users WHERE user_id = @user_id`, pgx.NamedArgs{
+		"user_id": myId,
+	}).Scan(&userWalletAddress, &userHandle)
+	if err != nil {
+		if errors.Is(err, pgx.ErrNoRows) {
+			return fiber.NewError(fiber.StatusInternalServerError, "Failed to get user wallet")
+		}
+		return err
+	}
+
+	mint, err := solana.PublicKeyFromBase58(mintString)
+	if err != nil {
+		return fmt.Errorf("Mint is invalid: %s", mintString)
+	}
+
+	var rewardStateAddress string
+	var minVotes int
+	err = app.pool.QueryRow(c.Context(), `SELECT reward_manager_state, min_votes FROM sol_reward_manager_inits WHERE mint = @mint`, pgx.NamedArgs{
+		"mint": mintString,
+	}).Scan(&rewardStateAddress, &minVotes)
+	if err != nil {
+		return fmt.Errorf("failed to get reward manager state: %w", err)
+	}
+	rewardManagerPubkey, err := solana.PublicKeyFromBase58(rewardStateAddress)
+	if err != nil {
+		return fmt.Errorf("failed to get init pubkey for reward manager state address: %w", err)
+	}
+
+	rewardManagerClient, err := reward_manager.NewRewardManagerClient(
+		app.solanaRpcClient,
+		app.solanaConfig.RewardManagerProgramID,
+		rewardManagerPubkey,
+		app.solanaConfig.RewardManagerLookupTable,
+		app.logger,
+	)
+
+	if err != nil {
+		return fmt.Errorf("failed to create reward manager client: %w", err)
+	}
+
+	bankAccount, err := app.claimableTokensClient.GetOrCreateUserBank(
+		c.Context(),
+		common.HexToAddress(userWalletAddress),
+		mint,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to get or create user bank: %w", err)
+	}
+
+	amount := uint64(0)
+	rewardAddress := ""
+	if redeemCode != "" {
+		// Read and burn code in one go
+		// Use CTE to capture old is_used value before updating
+		sql := `WITH old_row AS (
```

Comment:
Intentional choice to push this query as far down into the function as possible so that arbitrary errors in code that is not the actual claim process will just return early with 4xx/5xx. Once we get to this point, we're burning it and grabbing it in a single query (while also checking if it had already been used).
Note that using a code for the wrong mint and using a code that doesn't exist both return the same error.
If we want that to be different, will need to rework this.